### PR TITLE
Fine tuning launch.py

### PR DESCRIPTION
**Pull request by gongweibao**
1. Fine tuning launch.py.
2. Add unittest of it.

### CODE REVIEW

File: python/paddle/distributed/launch.py [hunk 1]
```diff
@@ -13,85 +13,42 @@
...
+
+def _print_arguments(args):
+    """Print argparse's arguments.
 
-        print("starting process ", i, entrypoint, entrypoint_args, curr_env)
-        fn = open("%s/workerlog.%d" % (log_dir, i), "w")
-        log_fns.append(fn)
-        cmd = [sys.executable, "-u", entrypoint] + entrypoint_args
-        procs.append(subprocess.Popen(cmd, stdout=fn, stderr=fn, env=curr_env))
+    Usage:
```
Comment (Yancey0623) [inline python/paddle/distributed/launch.py]: Usage? Better to show to how use the module `paddle.distributed.launch` ?

File: python/paddle/distributed/launch.py [hunk 2]
```diff
@@ -104,33 +61,142 @@ def parse_args():
...
+        "--print_config",
+        type=bool,
+        default=True,
+        help="Print the config or not")
+
+    parser.add_argument(
+        "--current_node_ip", type=str, help="The ip of current node.")
+
+    parser.add_argument(
+        "--nproc_per_node",
```
Comment (Yancey0623) [inline python/paddle/distributed/launch.py]: Can we use `nproc_per_node ` and `selected_gpus ` at the same time?

File: python/paddle/fluid/tests/unittests/test_launch.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+set -e
+
+node_ips="127.0.0.1"
+node_id="0"
+current_node_ip="127.0.0.1"
+
+distributed_args="--node_ips ${node_ips} --node_id ${node_id} --current_node_ip ${current_node_ip} --selected_gpus=0,1 --nproc_per_node=2"
+python -m paddle.distributed.launch ${distributed_args} multi_process.py
```
Comment (Yancey0623) [inline python/paddle/fluid/tests/unittests/test_launch.sh:12]: Assert something?

File: python/paddle/distributed/launch.py (hunk 1 above)
Comment (gongweibao) [reply python/paddle/distributed/launch.py]: Done. Thanks!

File: python/paddle/distributed/launch.py (hunk 2 above)
Comment (gongweibao) [reply python/paddle/distributed/launch.py]: Done. Use only one! Thanks!

File: python/paddle/fluid/tests/unittests/test_launch.sh (hunk 3 above)
Comment (gongweibao) [reply python/paddle/fluid/tests/unittests/test_launch.sh:12]: Done.

File: python/paddle/distributed/launch.py [hunk 4]
```diff
@@ -11,100 +11,54 @@
...
+
+Usage:
+    In both of single node training or multiple node training, this module 
+launch a process on each of the given gpu card.
+    
+    1. for single node four cards training
+       python -m torch.distributed.launch --selected_gpus="0,1,2,3,4" \
+         your_training_py (arg1 arg2 and all others)
+
+    2. for mulitple node training such as two node:192.168.0.16, 192.168.0.17
```
Comment (gavin1332) [inline python/paddle/distributed/launch.py]: the delimiter between ips is a comma AND a blank?

File: python/paddle/distributed/launch.py [hunk 5]
```diff
@@ -11,100 +11,54 @@
...
 # limitations under the License.
+"""
+paddle.distributed.launch is a module that spawns multiple distributed 
+process on each trainning node for gpu trainning.
+
+Usage:
+    In both of single node training or multiple node training, this module 
+launch a process on each of the given gpu card.
+    
+    1. for single node four cards training
```
Comment (gavin1332) [inline python/paddle/distributed/launch.py]: for single node training with four cards

File: python/paddle/distributed/launch.py [hunk 6]
```diff
@@ -11,100 +11,54 @@
...
+       python -m torch.distributed.launch --selected_gpus="0,1,2,3,4" \
```
Comment (gavin1332) [inline python/paddle/distributed/launch.py]: i think it better to indicate that "selected_gpus" option is alternative in this example, or show an example with minimum options explicitly.

Comment (gavin1332) [reply python/paddle/distributed/launch.py]: so as to illustrate the convenience of our toolkit

Comment (gongweibao) [reply python/paddle/distributed/launch.py]: Done.

File: python/paddle/distributed/launch.py (hunk 5 above)
Comment (gongweibao) [reply python/paddle/distributed/launch.py]: Done.Thanks!

File: python/paddle/distributed/launch.py [hunk 7]
```diff
@@ -11,100 +11,54 @@
...
+         your_training_py (arg1 arg2 and all others)
+
+    2. for mulitple node training such as two node:192.168.0.16, 192.168.0.17
+        on 192.168.0.16:
+            python -m torch.distributed.launch --node_ips="192.168.0.16, 192.168.0.17" \
```
Comment (gavin1332) [inline python/paddle/distributed/launch.py]: is it necessary to use a comma AND a blank as the delimiter between ips IN THIS EXAMPLE? Users may get confused about whether to follow this format strictly util they read the code

File: python/paddle/distributed/launch.py [hunk 8]
```diff
@@ -117,33 +71,128 @@ def parse_args():
...
+        default=True,
+        help="Print the config or not")
+
     parser.add_argument(
-        '--log_dir',
+        "--selected_gpus",
         type=str,
-        default="mylog",
-        help='directory to put logs per process.')
+        default="0,1,2,3,4,5,6,7",
```
Comment (gavin1332) [inline python/paddle/distributed/launch.py]: the default 8-cards configuration is not the minimum available set for some platform with only 4 cards. Can we detect this from CUDA_VISIBLE_DEVICES or by other ways

File: python/paddle/distributed/launch.py (hunk 7 above)
Comment (gongweibao) [reply python/paddle/distributed/launch.py]: It's a common choice to use a comma.
Done.

File: python/paddle/distributed/launch.py [hunk 9]
```diff
@@ -117,33 +71,128 @@ def parse_args():
...
+    """
+    procs = []
+    log_fns = []
+
+    default_env = os.environ.copy()
+
+    node_id = args.node_id
+    node_ips = [x.strip() for x in args.node_ips.split(',')]
+    assert node_id >= 0 and node_id < len(node_ips), \
+            "node_id:{} must be in range of the node_ips:{}".format(node_id, node_ips)
```
Comment (gavin1332) [inline python/paddle/distributed/launch.py]: "node_id:{} must be in range of the SIZE OF node_ips:{}".format(node_id, LEN(node_ips))

File: python/paddle/distributed/launch.py [hunk 10]
```diff
@@ -117,33 +71,128 @@ def parse_args():
...
+    num_nodes = len(node_ips)
+    selected_gpus = [x.strip() for x in args.selected_gpus.split(',')]
+    selected_gpus_num = len(selected_gpus)
+
+    trainers_endpoints = ""
+    for ip in node_ips:
+        for i in range(selected_gpus_num):
+            if trainers_endpoints != "":
+                trainers_endpoints += ","
+            trainers_endpoints += "%s:617%d" % (ip, i)
```
Comment (gavin1332) [inline python/paddle/distributed/launch.py:157]: make the port configurable, e.g. port_offset, default 6170

File: python/paddle/distributed/launch.py [hunk 11]
```diff
@@ -117,33 +71,128 @@ def parse_args():
...
+            "PADDLE_TRAINERS_NUM": "%d" % nranks,
+            "PADDLE_TRAINER_ENDPOINTS": trainers_endpoints
+        })
+
+        cmd = [sys.executable, "-u", args.training_script
+               ] + args.training_script_args
+
+        cmds.append(cmd)
+
+        if args.split_log_path is not None:
```
Comment (gavin1332) [inline python/paddle/distributed/launch.py]: useful argument, but the name is difficult to follow. my first reaction is change "/split/log/path" to ["split", "log", "path"].

Comment (gavin1332) [reply python/paddle/distributed/launch.py]: the original log_dir is just OK i think.

File: python/paddle/distributed/launch.py [hunk 12]
```diff
@@ -117,33 +71,128 @@ def parse_args():
...
+        type=bool,
+        default=True,
+        help="Print the config or not")
+
     parser.add_argument(
-        '--log_dir',
+        "--selected_gpus",
         type=str,
-        default="mylog",
-        help='directory to put logs per process.')
```
Comment (gavin1332) [inline python/paddle/distributed/launch.py:129]: more comments are required for describing the actions if this argument is not specified.

File: python/paddle/distributed/launch.py (hunk 10 above)
Comment (gongweibao) [reply python/paddle/distributed/launch.py:157]: Done.

**Review by gavin1332 — CHANGES_REQUESTED** [review]
useful upgrade! Cannot waiting for use. But some changes are required

File: python/paddle/distributed/launch.py (hunk 11 above)
Comment (gongweibao) [reply python/paddle/distributed/launch.py]: Done.

File: python/paddle/distributed/launch.py (hunk 9 above)
Comment (gongweibao) [reply python/paddle/distributed/launch.py]: Done.

File: python/paddle/distributed/launch.py (hunk 8 above)
Comment (gongweibao) [reply python/paddle/distributed/launch.py]: Done.

File: python/paddle/distributed/launch.py (hunk 7 above)
Comment (gavin1332) [reply python/paddle/distributed/launch.py]: yeah, i get the comma is necessary, but the blank is not.

File: python/paddle/distributed/launch.py (hunk 12 above)
Comment (gongweibao) [reply python/paddle/distributed/launch.py:129]: Done.Thanks!

**Review by gavin1332 — DISMISSED** [review]
LGTM

**Review by gavin1332 — APPROVED** [review]

**Pull request merged**